### PR TITLE
Add PIN Code support.

### PR DESCRIPTION
**Pull request by AlexProgrammerDE**
Change-type: minor
So this is our progress @tmigone @barhun :

- [x] Add Changetype in commit message (Added)
- [x] Add PIN Code request (Tested)
- [x] Delete bluetooth-agent file (Deleted)
- [x] Add character limit (Tested)
- [x] Fix Display error from character limit (Tested)
- [x] Fix security warning from bt-agent tool (Tested)
- [x] Fix useless chmod message. (Tested)
- [x] Add README info (Added)
- [x] Add Note that says that 16 characters are limit and they can be alphanumeric. (Added)
- [x] Replace the old example picture with a new with multiple variables set. (Added)
- [x] Delete useless dependencies (Deleted)
- [ ] Work on stability and troubleshooting (Waiting for test results on multiple devices.)

Bugs: https://github.com/balena-io-projects/balena-sound/pull/50#issuecomment-548922863

### CODE REVIEW

File: bluetooth-audio/start.sh [hunk 1]
```diff
@@ -26,7 +26,26 @@ sleep 2
...
+else
+if [[ -z "$BLUETOOTH_PIN_CODE" ]]; then
+bt-agent -c NoInputNoOutput &
+else
+cat <<EOF > /usr/src/bluetooth.cfg
+* $BLUETOOTH_PIN_CODE
+EOF
+bt-agent -p /usr/src/bluetooth.cfg &
+fi
+fi
```
Comment (AlexProgrammerDE) [inline bluetooth-audio/start.sh]: Sorry that it looks so confusing. EOF does not work when i use placeholder, so i maked it without them.

File: bluetooth-audio/start.sh [hunk 2]
```diff
@@ -26,7 +26,26 @@ sleep 2
...
-/usr/src/bluetooth-agent &
+hciconfig hci0 sspmode 0
+
+if [[ -n $(lshw | grep "Pi 2" -m 1) ]]; then
+if [[ -z "$BLUETOOTH_PIN_CODE" ]]; then
+BLUETOOTH_PIN_CODE=0000
+fi
+cat <<EOF > /usr/src/bluetooth.cfg
+* $BLUETOOTH_PIN_CODE
+EOF
```
Comment (barhun) [inline bluetooth-audio/start.sh]: Why not just using `echo "* $BLUETOOTH_PIN_CODE" > /usr/src/bluetooth.cfg`?

File: bluetooth-audio/start.sh [hunk 3]
```diff
@@ -26,7 +26,26 @@ sleep 2
 # can be confusing and it also hides those commands from the logs as well.
 printf "discoverable on\npairable on\nexit\n" | bluetoothctl > /dev/null
 
-/usr/src/bluetooth-agent &
+hciconfig hci0 sspmode 0
+
+if [[ -n $(lshw | grep "Pi 2" -m 1) ]]; then
+if [[ -z "$BLUETOOTH_PIN_CODE" ]]; then
+BLUETOOTH_PIN_CODE=0000
+fi
```
Comment (barhun) [inline bluetooth-audio/start.sh]: This could be replaced with `BLUETOOTH_PIN_CODE="${BLUETOOTH_PIN_CODE:-0000}"`

File: bluetooth-audio/start.sh [hunk 4]
```diff
@@ -26,7 +26,26 @@ sleep 2
...
+cat <<EOF > /usr/src/bluetooth.cfg
+* $BLUETOOTH_PIN_CODE
+EOF
+bt-agent -p /usr/src/bluetooth.cfg &
+else
+if [[ -z "$BLUETOOTH_PIN_CODE" ]]; then
+bt-agent -c NoInputNoOutput &
+else
+cat <<EOF > /usr/src/bluetooth.cfg
+* $BLUETOOTH_PIN_CODE
```
Comment (barhun) [inline bluetooth-audio/start.sh]: Why not just using echo "* $BLUETOOTH_PIN_CODE" > /usr/src/bluetooth.cfg?

**Comment by barhun** [comment]
By the way, I don't think the script `start.sh` needs to implement a solution customized for Raspberry Pi 2 as its default behavior would be the same with no customization. What I wanted to see was that we shouldn't ask for authentication in case of `BLUETOOTH_PIN_CODE` being unset, it now seems to have been implemented. Thanks for your efforts, @AlexProgrammerDE. (:

**Comment by barhun** [comment]
How about this simplified version of the same functionality?

```sh
if [ -z "$(lshw | grep 'Pi 2' -m 1)" -a -z "$BLUETOOTH_PIN_CODE" ]; then
  bt-agent -c NoInputNoOutput &
else
  echo "* ${BLUETOOTH_PIN_CODE:-0000}" > /usr/src/bluetooth.cfg
fi
```

**Comment by barhun** [comment]
What about this simplified version of the same functionality?

```sh
if [ -z "$(lshw | grep 'Pi 2' -m 1)" -a -z "$BLUETOOTH_PIN_CODE" ]; then
  bt-agent -c NoInputNoOutput &
else
  echo "* ${BLUETOOTH_PIN_CODE:-0000}" > /usr/src/bluetooth.cfg
fi
```

**Comment by AlexProgrammerDE** [comment]
Yeah looks good but i had seen on the RPI 2 unwanted results in the past. My Pi and your Pi (NoInputNoOutput)asked you for a Pin code which does not exist.

**Comment by barhun** [comment]
Well, I understand. This will cause inconsistent behavior on other devices without a built-in Bluetooth then, is there any way you know that we can detect if it is built-in or plugged-in?

**Comment by AlexProgrammerDE** [comment]
`-z "$(lshw | grep 'Pi 2' -m 1)"` was my solution for RPI 2 may be can we add in another way RPI 1

**Comment by barhun** [comment]
Could someone run `lsusb | grep -i bluetooth` on any device with built-in Bluetooth and tell us about the result?

**Comment by AlexProgrammerDE** [comment]
Bildschirmfoto von 2019-11-01 07-55-56

**Comment by barhun** [comment]
What about the output of `lsusb`?

**Comment by AlexProgrammerDE** [comment]
Bildschirmfoto von 2019-11-01 08-04-06

**Comment by AlexProgrammerDE** [comment]
This is on my RPI 2

**Comment by barhun** [comment]
Good, it seems that we can detect plugged-in Bluetooth devices? I still wonder what this command outputs in case of built-in Bluetooth. Can you test this on a rPi3/4? If it returns no output, then we can be said to have found the best solution. (:

**Comment by AlexProgrammerDE** [comment]
Sorry i do not have a RPI 3 or 4.

**Comment by AlexProgrammerDE** [comment]
Look at the first reply at this website please: https://discourse.osmc.tv/t/i-have-rpi-3-model-b-rev-1-2-and-couple-of-questions/39754

**Comment by barhun** [comment]
I think this'd be better to use if it works out with rPi3/4 than singling out the case of rPi2, so we need the help of others with those devices. (:

**Comment by AlexProgrammerDE** [comment]
So i thinked about it... when there is a Bluetooth dongle happens this weird behaviour.
It could be a good way to say Bluetooth is build in when no USB dongle is connected and when a Bluetooth dongle is connected it is plugged in:

if [ -z "$(lsusb | grep -i bluetooth)" -a -z "$BLUETOOTH_PIN_CODE" ]; then
  bt-agent -c NoInputNoOutput &
else
  echo "* ${BLUETOOTH_PIN_CODE:-0000}" > /usr/src/bluetooth.cfg
  bt-agent -p /usr/src/bluetooth.cfg &
fi

**Comment by barhun** [comment]
Looks good to me.

**Comment by AlexProgrammerDE** [comment]
So i will push it. Could you please give me a little tutorial how to overwrite it?

**Comment by barhun** [comment]
https://github.com/wprig/wprig/wiki/How-to-squash-commits

**Comment by AlexProgrammerDE** [comment]
So i need to be in my fork root directory?

**Comment by barhun** [comment]
Yes, do all the changes you want and commit them. You can squash many commits into one as explained in the article above and then push your commits forcing the remote to be the same with your local history.

**Comment by AlexProgrammerDE** [comment]
@barhun So i hope the checks will pass. I put everything now to one commit and put the right text into the description.

**Comment by AlexProgrammerDE** [comment]
Maybe works it with closing and reopening.

**Comment by AlexProgrammerDE** [comment]
@barhun this works well i can trigger new checks with closing and opening the PR's

**Comment by AlexProgrammerDE** [comment]
Yeah i will fix it.

**Comment by AlexProgrammerDE** [comment]
So reopen works well. @barhun i found a easyer way:

git add .
then:
git commit --amend
and then force push. :+1:

**Comment by AlexProgrammerDE** [comment]
Should i write a README info?

**Comment by barhun** [comment]
Hey, @chrisys... We've got some questions! (:

**Comment by AlexProgrammerDE** [comment]
So... i found a little bug. When we use more than 16 characters as the Pin Code will my phone prompt me before i could write my 17th character that more than 16 characters are impossible. So could we add a reset to 0000 when more than 16 characters are used.

**Comment by AlexProgrammerDE** [comment]
@tmigone should i add something before you approve?

**Comment by tmigone** [comment]
No, the PR looks good. I only need to finish testing and we are good (hopefully today!).

**Comment by AlexProgrammerDE** [comment]
Thanks.

**Comment by tmigone** [comment]
Hey I've been testing on a Pi3 and i have mixed results unfortunately. 
If I add a custom PIN CODE it seems to work, though not as reliably as you would expect (i had some disconnections and false connections too).
Without the custom PIN CODE (if i don't add the ENV var) its hit or miss. 

I'll be testing on more devices, and with RPi2 / RPi4 and get back to you. 
Definitely we need to prioritize RPi3 and RPi4, not only because they are newer boards but because of the built-in bluetooth (harder to support usb bt dongles).

We'll get it to work though!

**Comment by AlexProgrammerDE** [comment]
> Hey I've been testing on a Pi3 and i have mixed results unfortunately.
> If I add a custom PIN CODE it seems to work, though not as reliably as you would expect (i had some disconnections and false connections too).
> Without the custom PIN CODE (if i don't add the ENV var) its hit or miss.
> 
> I'll be testing on more devices, and with RPi2 / RPi4 and get back to you.
> Definitely we need to prioritize RPi3 and RPi4, not only because they are newer boards but because of the built-in bluetooth (harder to support usb bt dongles).
> 
> We'll get it to work though!

So ok. I will work on it tomorrow. I had the same result with the Variable set. I solved the problem with going nearer to the device. I think this cant be solved, because it is maybe a hardware problem. The built-in bluetooth is too weak to send and request Pin codes over the distance. But after pairing should it work like before this PR.

**Comment by AlexProgrammerDE** [comment]
By the way what did you mean with:

>  its hit or miss.

**Comment by tmigone** [comment]
Sometimes it would connect right away while other times it would not connect at all. 
It needs more testing from my part, trying to get a Pi2 at the moment.

**Comment by AlexProgrammerDE** [comment]
Ok

**Comment by AlexProgrammerDE** [comment]
Sorry. I readed more of the page and saw that in this constellation it means that both of them need to be true

**Comment by AlexProgrammerDE** [comment]
I will add code which deletes this warning:

Bildschirmfoto von 2019-11-02 09-42-38

**Comment by AlexProgrammerDE** [comment]
So now i am added the 16 character limit. If more than 16 character are used the RPI uses 0000 as the PIN code. I will push in a few seconds.

**Comment by AlexProgrammerDE** [comment]
So now will i push code which fixes a display error in the dashboard because printf did not create a newline at the end of the line.

**Comment by AlexProgrammerDE** [comment]
PS: I did add checkboxes at the top which will be our progress. :smile_cat:

**Comment by AlexProgrammerDE** [comment]
So now i added a README info.

**Comment by AlexProgrammerDE** [comment]
@barhun  @tmigone how does it look now?

**Comment by AlexProgrammerDE** [comment]
So now i have overwritten the device-config-name.png file with another picture from me with more variables set. This can help people which want to set multiple variables.

**Comment by AlexProgrammerDE** [comment]
I will now write a README edit that informs that 16 characters are the maximum and the characters can be alphanumeric.

**Comment by AlexProgrammerDE** [comment]
Could someone of you review my changes? @barhun @tmigone

**Comment by AlexProgrammerDE** [comment]
Do you have got results @tmigone ?

**Comment by AlexProgrammerDE** [comment]
Now i added a example pin code variable picture and restored the old picture.

**Comment by AlexProgrammerDE** [comment]
Now i deleted the python dependencies.

**Comment by tmigone** [comment]
Hey @AlexProgrammerDE @barhun, I took the time to do some more testing and the results were mixed as I reported last week. I tested with RPi3 and RPi4 and the connection process was very unreliable, sometimes it would connect right away without asking pin code, then disconnect and then ask for the code; some other times it would fail to connect even with the correct pin code. 

I went back to the original code (using the python bluetooth agent) and made some changes so that it could support the pin code functionality. You can check the changes on PR [#51](https://github.com/balena-io-projects/balena-sound/pull/51) and also on [this](https://github.com/balena-io-projects/balena-sound/tree/20-add-pin-code) branch. 
I tested this on RPi3 and RPi4 and the connection process is a lot more smoother, both with and without pin code. I would appreciate it if you could test this on the Raspberry Pi 2 since my bluetooth dongle seems to no longer work.

Once again I wanted to thank you both for the work and contributions you've made to get this feature to work.

**Comment by AlexProgrammerDE** [comment]
@tmigone found a solution for it. Closing.

**Pull request closed without merging**